# Basics of MLP

- Objective: create vanilla neural networks (i.e., Multilayer perceptrons) for simple regression/classification tasks with Keras

## MLP Structures
- Each MLP model is consisted of one input layer, several hidden layers, and one output layer
- Number of neurons in each layer is not limited
<img src="http://cs231n.github.io/assets/nn1/neural_net.jpeg" style="width: 500px"/>
<br>
<center>**MLP with one hidden layer**</center>
- Number of input neurons: 3
- Number of hidden neurons: 4
- Number of output neurons: 2


<img src="http://cs231n.github.io/assets/nn1/neural_net2.jpeg" style="width: 500px"/>
<br>
<center>**MLP with two hidden layers**</center>

- Number of input neurons: 3
- Number of hidden neurons: (4,4)
- Number of output neurons: 2

## MLP for Regression tasks

- When the target (y) is continuous (real)
- For loss function and evaluation metric, mean squared error (MSE) is commonly used

In [2]:
from keras.datasets import boston_housing

Using TensorFlow backend.


In [10]:
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()

In [12]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(404, 13)
(102, 13)
(404,)
(102,)


## Dataset Description
- Boston housing dataset has total 506 data instances (404 training & 102 test)
- 13 attributes (features) to predict "the median values of the houses at a location"
- Doc: https://keras.io/datasets/

In [16]:
from keras.layers import Activation, Dense
from keras.models import Sequential
model = Sequential()

## 1-1. Adding layers
- Keras layers can be added to the model
- Adding layers are like stacking lego blocks one by one
- Doc: https://keras.io/layers/core/

In [17]:
# Keras model with two hidden layer with 10 neurons each 
model.add(Dense(10, input_shape = (13,)))    # Input layer => input_shape should be explicitly designated
model.add(Activation('sigmoid'))
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(Activation('sigmoid'))
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(Activation('sigmoid'))
model.add(Dense(1))                          # Output layer => output dimension = 1 since it is regression problem

In [18]:
# This is equivalent to the above code block
model.add(Dense(10, input_shape = (13,), activation = 'sigmoid'))
model.add(Dense(10, activation = 'sigmoid'))
model.add(Dense(10, activation = 'sigmoid'))
model.add(Dense(1))

## 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
- Doc (optimizers): https://keras.io/optimizers/
- Doc (losses): https://keras.io/losses/

In [19]:
from keras import optimizers

In [20]:
sgd = optimizers.SGD(lr = 0.01)    # stochastic gradient descent optimizer

In [21]:
model.compile(optimizer = sgd, loss = 'mean_squared_error', metrics = ['mse'])    # for regression problems, mean squared error (MSE) is often employed

## Summary of the model

In [22]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 10)                140       
_________________________________________________________________
activation_1 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                110       
_________________________________________________________________
activation_2 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                110       
_________________________________________________________________
activation_3 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_4 (Dense)              (None, 1)                

## 2. Training

- Training the model with training data provided

In [24]:
model.fit(x_train, y_train, batch_size = 50, epochs = 100, verbose = 1)

Epoch 1/100
404/404 [==============================] - 1s 2ms/step - loss: 327.9146 - mse: 327.9146
Epoch 2/100
404/404 [==============================] - 0s 52us/step - loss: 95.0720 - mse: 95.0720
Epoch 3/100
404/404 [==============================] - 0s 47us/step - loss: 84.8646 - mse: 84.8645
Epoch 4/100
404/404 [==============================] - 0s 45us/step - loss: 85.3727 - mse: 85.3727
Epoch 5/100
404/404 [==============================] - 0s 42us/step - loss: 85.0978 - mse: 85.0978
Epoch 6/100
404/404 [==============================] - 0s 42us/step - loss: 84.8347 - mse: 84.8347
Epoch 7/100
404/404 [==============================] - 0s 45us/step - loss: 85.8435 - mse: 85.8435
Epoch 8/100
404/404 [==============================] - 0s 42us/step - loss: 85.1108 - mse: 85.1108
Epoch 9/100
404/404 [==============================] - 0s 40us/step - loss: 85.3562 - mse: 85.3562
Epoch 10/100
404/404 [==============================] - 0s 40us/step - loss: 85.7313 - mse: 85.7313
Epoch 11

404/404 [==============================] - 0s 57us/step - loss: 84.9835 - mse: 84.9835
Epoch 83/100
404/404 [==============================] - 0s 50us/step - loss: 85.6016 - mse: 85.6016
Epoch 84/100
404/404 [==============================] - 0s 54us/step - loss: 86.2915 - mse: 86.2915
Epoch 85/100
404/404 [==============================] - 0s 57us/step - loss: 84.6994 - mse: 84.6994
Epoch 86/100
404/404 [==============================] - 0s 64us/step - loss: 85.0582 - mse: 85.0582
Epoch 87/100
404/404 [==============================] - 0s 59us/step - loss: 84.9033 - mse: 84.9033
Epoch 88/100
404/404 [==============================] - 0s 42us/step - loss: 85.3963 - mse: 85.3963
Epoch 89/100
404/404 [==============================] - 0s 42us/step - loss: 86.1413 - mse: 86.1413
Epoch 90/100
404/404 [==============================] - 0s 40us/step - loss: 85.1318 - mse: 85.1318
Epoch 91/100
404/404 [==============================] - 0s 40us/step - loss: 84.6707 - mse: 84.6707
Epoch 92/100


## 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
- Doc (metrics): https://keras.io/metrics/

In [26]:
results = model.evaluate(x_test, y_test)

102/102 [==============================] - 0s 1ms/step


In [27]:
print(model.metrics_names)     # list of metric names the model is employing
print(results)                 # actual figure of metrics computed

['loss', 'mse']
[83.83948501886105, 83.8394775390625]


In [28]:
print('loss: ', results[0])
print('mse: ', results[1])

loss:  83.83948501886105
mse:  83.8394775390625


## MLP for classification tasks

- When the target (y) is discrete (categorical)
- For loss function, cross-entropy is used and for evaluation metric, accuracy is commonly used

In [29]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

In [30]:
whole_data = load_breast_cancer()

In [33]:
whole_data

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [34]:
x_data = whole_data.data
y_data = whole_data.target

In [35]:
x_data

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [36]:
y_data

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [38]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(398, 30)
(171, 30)
(398,)
(171,)


## Dataset Description
- Breast cancer dataset has total 569 data instances (212 malign, 357 benign instances)
- 30 attributes (features) to predict the binary class (M/B)
- Doc: http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer

In [37]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size = 0.3, random_state = 7)

# 1. Creating a model
- Same with regression model at the outset

In [39]:
from keras.models import Sequential

In [40]:
model = Sequential()

# 1-1. Adding layers
- Keras layers can be added to the model
- Adding layers are like stacking lego blocks one by one
- It should be noted that as this is a classification problem, sigmoid layer (softmax for multi-class problems) should be added
- Doc: https://keras.io/layers/core/

In [41]:
# Keras model with two hidden layer with 10 neurons each 
model.add(Dense(10, input_shape = (30,)))    # Input layer => input_shape should be explicitly designated
model.add(Activation('sigmoid'))
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(Activation('sigmoid'))
model.add(Dense(10))                         # Hidden layer => only output dimension should be designated
model.add(Activation('sigmoid'))
model.add(Dense(1))                          # Output layer => output dimension = 1 since it is regression problem
model.add(Activation('sigmoid'))

In [43]:
# This is equivalent to the above code block
model.add(Dense(10, input_shape = (13,), activation = 'sigmoid'))
model.add(Dense(10, activation = 'sigmoid'))
model.add(Dense(10, activation = 'sigmoid'))
model.add(Dense(1, activation = 'sigmoid'))

# 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
- Doc (optimizers): https://keras.io/optimizers/
- Doc (losses): https://keras.io/losses/

In [45]:
from keras import optimizers

In [46]:
sgd = optimizers.SGD(lr = 0.01)    # stochastic gradient descent optimizer

In [47]:
model.compile(optimizer = sgd, loss = 'binary_crossentropy', metrics = ['accuracy'])

## Summary of the model

In [48]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 10)                310       
_________________________________________________________________
activation_4 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_10 (Dense)             (None, 10)                110       
_________________________________________________________________
activation_5 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_11 (Dense)             (None, 10)                110       
_________________________________________________________________
activation_6 (Activation)    (None, 10)                0         
_________________________________________________________________
dense_12 (Dense)             (None, 1)                

## 2. Training
- Training the model with training data provided

In [50]:
model.fit(x_train, y_train, batch_size = 50, epochs = 100, verbose = 1)

Epoch 1/100
398/398 [==============================] - 1s 1ms/step - loss: 0.6833 - accuracy: 0.6055
Epoch 2/100
398/398 [==============================] - 0s 58us/step - loss: 0.6814 - accuracy: 0.6055
Epoch 3/100
398/398 [==============================] - 0s 50us/step - loss: 0.6797 - accuracy: 0.6055
Epoch 4/100
398/398 [==============================] - 0s 48us/step - loss: 0.6784 - accuracy: 0.6055
Epoch 5/100
398/398 [==============================] - 0s 48us/step - loss: 0.6773 - accuracy: 0.6055
Epoch 6/100
398/398 [==============================] - 0s 45us/step - loss: 0.6762 - accuracy: 0.6055
Epoch 7/100
398/398 [==============================] - 0s 48us/step - loss: 0.6754 - accuracy: 0.6055
Epoch 8/100
398/398 [==============================] - 0s 45us/step - loss: 0.6746 - accuracy: 0.6055
Epoch 9/100
398/398 [==============================] - 0s 45us/step - loss: 0.6740 - accuracy: 0.6055
Epoch 10/100
398/398 [==============================] - 0s 45us/step - loss: 0.6735

398/398 [==============================] - 0s 50us/step - loss: 0.6707 - accuracy: 0.6055
Epoch 80/100
398/398 [==============================] - 0s 48us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 81/100
398/398 [==============================] - 0s 48us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 82/100
398/398 [==============================] - 0s 45us/step - loss: 0.6707 - accuracy: 0.6055
Epoch 83/100
398/398 [==============================] - 0s 43us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 84/100
398/398 [==============================] - 0s 43us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 85/100
398/398 [==============================] - 0s 45us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 86/100
398/398 [==============================] - 0s 50us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 87/100
398/398 [==============================] - 0s 60us/step - loss: 0.6708 - accuracy: 0.6055
Epoch 88/100
398/398 [==============================] - 0s 50us/step - loss: 0.6708 - 

## 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
- Doc (metrics): https://keras.io/metrics/

In [51]:
results = model.evaluate(x_test, y_test)

171/171 [==============================] - 0s 472us/step


In [52]:

print(model.metrics_names)     # list of metric names the model is employing
print(results)                 # actual figure of metrics computed

['loss', 'accuracy']
[0.6394409505944503, 0.6783625483512878]


In [53]:
print('loss: ', results[0])
print('accuracy: ', results[1])

loss:  0.6394409505944503
accuracy:  0.6783625483512878
